backup: revalidate restored indexes if added during inc backup

dt opened the pull request:
Incremental backups of indexes being built can miss some index entries.

Recent changes mean backup now avoids backing up such an index until it
is complete, but older backups may have these indexes that were caught
mid-backfill. This change identifies indexes that were offline / building
in any layer of the backup and then for those that are to be restored
to a public state, re-runs validation after restore ensure the index
has the expected entries.

Currently if any restored index is missing entries, it is marked offline for
reconstruction post-RESTORE.

Release note (sql change): RESTORE now re-validates restored indexes if they were restored
from an incremental backup that was taken while the index was being created.

dt committed:
backup: revalidate restored indexes if added during inc backup

Incremental backups of indexes being built can miss some index entries.

Recent changes mean backup now avoids backing up such an index until it
is complete, but older backups may have these indexes that were caught
mid-backfill. This change identifies indexes that were offline / building
in any layer of the backup and then for those that are to be restored
to a public state, re-runs validation after restore ensure the index
has the expected entries.

If any restored index is missing entries, the restore marks it as adding
rather than public, so that a job can be synthesizes to rebuild it.

Release note (sql change): RESTORE now re-validates restored indexes if they were restored
from an incremental backup that was taken while the index was being created.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62701)
<!-- Reviewable:end -->

dt commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -1696,6 +1770,15 @@ func (r *restoreResumer) publishDescriptors(
+			}
+			mutTable.RemovePublicNonPrimaryIndex(found.Ordinal())
+			mutTable.AddIndexMutation(found.IndexDesc(), descpb.DescriptorMutation_ADD)
+		}
```
@ajwerner I occasionally see ` duplicate: index "balance_idx_4" in the middle of being added, not yet public` when I get to the write below. Any obvious mistake here?

dt replied:
huh, something is weird. I added printlns of AllNonDropIndexes before and after the remove/add pair and see this: 

```
tbl 54 idx 1 "primary" (adding: false)
tbl 54 idx 2 "balance_idx_1" (adding: false)
tbl 54 idx 3 "balance_idx_2" (adding: false)
tbl 54 idx 4 "balance_idx_3" (adding: false)
tbl 54 idx 5 "balance_idx_4" (adding: false)
removing idx 4 (in pos 3) from tbl 54 and re-adding as "balance_idx_3"
 -> tbl 54 idx 1 "primary" (adding: false)
 -> tbl 54 idx 2 "balance_idx_1" (adding: false)
 -> tbl 54 idx 3 "balance_idx_2" (adding: false)
 -> tbl 54 idx 5 "balance_idx_4" (adding: false)
 -> tbl 54 idx 5 "balance_idx_4" (adding: true)
```
that is unexpected -- we moved 3 (idx_4) and re-added it, but we see 5 duplicated instead

ajwerner reviewed (commented):
What's the status of the testing here? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62701#-:-MXb5p5E25DawvDBakFs:b-rixosb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)



<!-- Sent from Reviewable.io -->

dt commented:
The added test fails (observes incorrect index count) without the associated code change, and passes with it. However once rebased on #62572, the test will still pass, but would no longer fail without the code change. We’ll need to vendor a bad backup since we will no longer be making them. But pre-rebase, the test does what I want, so I think it does work.

dt commented:
Cluster restore still needs attention since we’ll need to do job synthesis we usually don’t need there though

dt commented:
Okay, updated this with some vendored 20.2 cluster backups created prior to #62572 merging. Also added test coverage for cluster restore and inverted indexes.

shermanCRL commented:
Addresses parent issue #63162

pbardea commented on pkg/ccl/backupccl/restore_old_versions_test.go:
```diff
@@ -248,3 +250,194 @@ func restoreOldVersionClusterTest(exportDir string) func(t *testing.T) {
+					}
+				case 0:
+					found := false
+					for i := 0; i < 1; i++ {
```
Why do we only need 1 attempt here? Is there a difference here b/w JSON and normal indexes, or do we just do the waiting in the previous case?

dt replied:
Yep, I think the previous case saves us. upped it to be 50 too just for consistency.

pbardea commented on pkg/ccl/backupccl/restore_old_versions_test.go:
```diff
@@ -248,3 +250,194 @@ func restoreOldVersionClusterTest(exportDir string) func(t *testing.T) {
+					if res != expected {
+						t.Fatalf("got %d, expected %d", res, expected)
+					}
+				case 0:
```
Was the exclusion of `case 1`/geo indexes intentional? I don't think I follow why those are different.

dt replied:
Added a comment

pbardea commented on pkg/ccl/backupccl/restore_old_versions_test.go:
```diff
@@ -248,3 +250,194 @@ func restoreOldVersionClusterTest(exportDir string) func(t *testing.T) {
+		))
+		backupDirs = append(backupDirs, `'nodelocal://0/final'`)
+	}
+	os.Rename(dir, path/to/testdata)
```
nit if you're going to push again: although this will fail anyway, I think this is missing a closing brace.

dt replied:
heh, indeed. done.

pbardea reviewed (approved):
Nonblocking comments. Mostly questions.

dt commented:
TFTR!

bors r=pbardea
